fdbserver/BlobManifest.actor.cpp

```diff
@@ -0,0 +1,369 @@
+/*
+ * BlobManifest.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/BackupContainer.h"
+#include "fdbserver/Knobs.h"
+#include "flow/FastRef.h"
+#include "flow/flow.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/BlobConnectionProvider.h"
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/KeyRangeMap.h"
+#include "fdbclient/SystemData.h"
+#include "fdbclient/BackupContainerFileSystem.h"
+#include "fdbclient/BlobGranuleReader.actor.h"
+#include "fdbserver/BlobGranuleServerCommon.actor.h"
+
+#include "flow/actorcompiler.h" // has to be last include
+#include "fmt/core.h"
+
+//
+// This module offers routines to dump or load blob manifest file, which is used for full restore from granules
+//
+
+static std::string MANIFEST_FILENAME = "manifest"; // Default manifest file name on external blob storage
+
+#define ENABLE_DEBUG_PRINT true
+template <typename... T>
+inline void dprint(fmt::format_string<T...> fmt, T&&... args) {
+	if (ENABLE_DEBUG_PRINT)
+		fmt::print(fmt, std::forward<T>(args)...);
+}
+
+// This class dumps blob manifest to external blob storage.
+class BlobManifestDumper : public ReferenceCounted<BlobManifestDumper> {
+public:
+	BlobManifestDumper(Database& db, Reference<BlobConnectionProvider> blobConn) : db_(db), blobConn_(blobConn) {}
+	virtual ~BlobManifestDumper() {}
+
+	// Execute the dumper
+	ACTOR static Future<Void> execute(Reference<BlobManifestDumper> self) {
+		try {
+			state Standalone<BlobManifest> manifest;
+			Standalone<VectorRef<KeyValueRef>> rows = wait(getSystemKeys(self));
+			manifest.rows = rows;
+			Value data = encode(manifest);
+			wait(writeToFile(self, data));
+		} catch (Error& e) {
+			dprint("WARNING: unexpected blob manifest dumper error {}\n", e.what()); // skip error handling for now
+		}
+		return Void();
+	}
+
+private:
+	// Return system keys that to be backed up
+	ACTOR static Future<Standalone<VectorRef<KeyValueRef>>> getSystemKeys(Reference<BlobManifestDumper> self) {
+		state Standalone<VectorRef<KeyValueRef>> rows;
+		loop {
+			state Transaction tr(self->db_);
```

Comment:
define the transaction outside the transaction retry loop because onError() has an exponential backoff that will not work when the transaction is recreated on each failure.

also add PRIORITY_SYSTEM_IMMEDIATE and LOCK_AWARE so that this transaction will run even if the cluster is throttled by ratekeeper, and even when the cluster is locked